src/qmflows/settings.py

```diff
@@ -96,7 +102,59 @@ def update(self, other):
                     pred1 = (br in other[name] and other[name][br])
                     pred2 = (br in self[name] and self[name][br])
                     if pred1 or pred2:
-                        self[name] = Settings()
+                        self[name] = cls()
                     self[name].update(other[name])
             else:
                 self[name] = other[name]
+
+
+def _unsuported_operation(meth: Callable) -> Callable[..., NoReturn]:
+    """Decorate a method such that it raises a :exc:`TypeError`."""
+    @wraps(meth)
+    def new_meth(self, *args, **kwargs) -> NoReturn:
+        raise TypeError(f"{self.__class__.__name__!r} does not "
+                        "support item assignment or deletion")
+    return new_meth
+
+
+def _super_method(meth: Callable[..., '_Settings']) -> Callable[..., '_Settings']:
+    """Decorate a method such that it's body is executed by the super-class."""
+    @wraps(meth)
+    def new_meth(self, *args, **kwargs) -> '_Settings':
+        cls = type(self)
+        func = getattr(super(), meth.__name__)
+        ret = func(*args, **kwargs)
+        return cls(ret)
+    return new_meth
+
+
+class _Settings(Settings):
+    """Immutable-ish counterpart of :class:`Settings`."""
+
+    def __init__(self, *args, **kwargs):
+        cls = type(self)
+        set_item = super().__setitem__
+
+        dict.__init__(self, *args, **kwargs)
+        for k, v in self.items():
+            if isinstance(v, dict):
+                set_item(k, cls(v))
+            elif isinstance(v, list):
+                set_item(k, [cls(i) if isinstance(i, dict) else i for i in v])
+
+    def __missing__(self, name: str) -> NoReturn:
+        raise KeyError(name)
+
+    __setitem__ = _unsuported_operation(Settings.__setitem__)
+    __delitem__ = _unsuported_operation(Settings.__delitem__)
+    clear = _unsuported_operation(Settings.clear)
+    popitem = _unsuported_operation(Settings.popitem)
+    setdefault = _unsuported_operation(Settings.setdefault)
+    update = _unsuported_operation(Settings.update)
+
+    overlay = _super_method(Settings.overlay)
```

Comment:
what is the difference in behaviour between `overlay = _super_method(Settings.overlay)` and `overlapy = Setting.overlay`? I think that Python should take care of generating the right class instance, otherwise it is a bit of rewriting the polymorphic behaviour of Python

Reply:
These methods decorated by ``_super_method()`` all rely on the ``_Setings.__setitem__()`` operation in one way or another. The thing is... ``_Setings.__setitem__()`` raise a ``TypeError`` by design.

So, how do we get around this issue? By executing the method in the (mutable) superclass and then converting the output into the proper (immutable) type afterwards.

Reply:
I didn't know about the `TypeError` in that case your solution is very good and quite elegant 😊 I think you are doing a great job!